cmd/easeprobe/main.go

```diff
@@ -45,48 +48,48 @@ func main() {
 	yamlFile := flag.String("f", getEnvOrDefault("PROBE_CONFIG", "config.yaml"), "configuration file")
 	flag.Parse()
 
-	conf, err := conf.New(yamlFile)
+	c, err := conf.New(yamlFile)
 	if err != nil {
 		log.Fatalln("Fatal: Cannot read the YAML configuration file!")
 		os.Exit(-1)
 	}
-	defer conf.CloseLogFile()
+	defer c.CloseLogFile()
 
 	// if dry notification mode is specificed in command line, overwrite the configuration
 	if *dryNotify {
-		conf.Settings.Notify.Dry = *dryNotify
+		c.Settings.Notify.Dry = *dryNotify
 	}
 
-	if conf.Settings.Notify.Dry {
+	if c.Settings.Notify.Dry {
 		log.Infoln("Dry Notification Mode...")
 	}
 
 	// Probers
-	probers := conf.AllProbers()
+	probers := c.AllProbers()
 
 	// Notification
-	notifies := conf.AllNotifiers()
-
-	done := make(chan bool)
-	run(probers, notifies, done)
-
-}
-
-// 1) all of probers send the result to notify channel
-// 2) go through all of notification to notify the result.
-// 3) send the SLA report
-func run(probers []probe.Prober, notifies []notify.Notify, done chan bool) {
-
-	dryNotify := conf.Get().Settings.Notify.Dry
+	notifies := c.AllNotifiers()
 
+	// wait group for probers
+	var wg sync.WaitGroup
+	// the exit channel for all probers
+	doneProbe := make(chan bool)
```

Comment:
I think the buffered chan is better than the unbuffered chan in the scenarios because the unbuffered chan will block go routine who writes data to it when no goroutine receives data on it. (I think the probe's goroutine may exit abnormally due to a runtime panic or bug). So I recommend allocating a buffered with a size of probers

Reply:
good comments!